File: dotnet/DesignPatterns/BehavioralDesignPatterns/StateDesignPattern/StateDesignPattern.Example1/PendingState.cs
namespace StateDesignPattern.Example1;

/// <summary>
/// Concrete State - represents a pending order waiting to be processed.
/// </summary>
public class PendingState : IOrderState
{
    public void ProcessOrder(Order order)
    {
        Console.WriteLine("Order is being processed...");
        order.SetState(new ProcessingState());
        Console.WriteLine($"Order state changed to: {order.GetCurrentStateName()}");
    }

    public void ShipOrder(Order order)
    {
        Console.WriteLine("Cannot ship. Order must be processed first.");
    }

    public void DeliverOrder(Order order)
    {
        Console.WriteLine("Cannot deliver. Order must be shipped first.");
    }

    public void CancelOrder(Order order)
    {
        Console.WriteLine("Order cancelled successfully.");
        order.SetState(new CancelledState());
        Console.WriteLine($"Order state changed to: {order.GetCurrentStateName()}");
    }

    public string GetStateName() => "Pending";
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/StateDesignPattern/StateDesignPattern.Example1/src/PendingState.ts
import { IOrderState } from './IOrderState';
import { Order } from './Order';
import { ProcessingState } from './ProcessingState';
import { CancelledState } from './CancelledState';

export class PendingState implements IOrderState {
  processOrder(order: Order): void {
    console.log('Order is being processed...');
    order.setState(new ProcessingState());
    console.log(`Order state changed to: ${order.getCurrentStateName()}`);
  }

  shipOrder(order: Order): void {
    console.log('Cannot ship. Order must be processed first.');
  }

  deliverOrder(order: Order): void {
    console.log('Cannot deliver. Order must be shipped first.');
  }

  cancelOrder(order: Order): void {
    console.log('Order cancelled successfully.');
    order.setState(new CancelledState());
    console.log(`Order state changed to: ${order.getCurrentStateName()}`);
  }

  getStateName(): string {
    return 'Pending';
  }
}
